fabtests/rdm_atomic, ubertest: move atomic verification to common code and use in functional test

aingerson opened the pull request:
Ubertest copies the software implementation of atomics from libfabric in order to add atomic operation data validation. This is done by saving the sent data and performing a local atomic in order to validate against the actual data calculated by the libfabric provider.
This patch set moves that implementation from ubertest to the common code for other tests (specifically the fi_rdm_atomic test) to leverage it.
Ubertest atomics with validation is not run in the CI by default but he rdm_atomic test is. Because of the lack of validation in the functional fabtest, an incorrect fetched buffer pointer was never caught. This patch set also enables running rdm_atomic with validation in the standard and short test sets in runfabtests.sh so we can catch this in the future.
During testing of these changes, it properly caught the previously missed bug and properly passes when it was fixed.

aingerson commented on fabtests/common/shared.c:
```diff
@@ -3653,22 +3698,76 @@ int ft_check_atomic(enum ft_atomic_opcodes atomic, enum fi_op op,
+
+		check_buf = dev_host_buf;
+
+		ret = ft_hmem_copy_from(opts.iface, opts.device, dev_host_comp,
+					cmp, count * datatype_to_size(type));
+		if (ret) {
+			FT_ERR("Failed to copy from atomic buffer\n");
+			return ret;
+		}
+
+		check_comp = dev_host_comp;
```
No, I think I left it because coverity gets really confused by the unused variables with the atomic cases. But I'm going to refactor this whole check which is looking really messy. I'll push an update when I figure out the other failures.

j-xiong commented:
Looks like MS compiler doesn't like the complex types:
```
 dgram_pingpong.c
C:\projects\libfabric\fabtests\include\ofi_atomic.h(44): error C2061: syntax error: identifier 'ofi_complex_float' [C:\projects\libfabric\fabtests\fabtests.vcxproj]
C:\projects\libfabric\fabtests\include\ofi_atomic.h(44): error C2059: syntax error: ';' [C:\projects\libfabric\fabtests\fabtests.vcxproj]
C:\projects\libfabric\fabtests\include\ofi_atomic.h(45): error C2061: syntax error: identifier 'ofi_complex_double' [C:\projects\libfabric\fabtests\fabtests.vcxproj]
C:\projects\libfabric\fabtests\include\ofi_atomic.h(45): error C2059: syntax error: ';' [C:\projects\libfabric\fabtests\fabtests.vcxproj]
C:\projects\libfabric\fabtests\include\ofi_atomic.h(46): error C2061: syntax error: identifier 'ofi_complex_long_double' [C:\projects\libfabric\fabtests\fabtests.vcxproj]
C:\projects\libfabric\fabtests\include\ofi_atomic.h(46): error C2059: syntax error: ';' 
```

aingerson commented:
@acgoldma This PR is adding atomic verification to the regular fabtests runs. psm3 was failing a lot of them so I've added a few patches to fix most of them. The last patch disables the remaining failing combinations instead of fixing them because I couldn't completely root cause - long double and complex long double reads/swaps are failing and I'm not sure why. Switching the sets to a memcpy of the datatype size also fixes the issue but it takes a performance hit (and also affects all other data types for those ops) so I opted to disable instead of adding that fix. I did this by having all the reads and compares go through the PSM_CUDA | PSM_ONEAPI implementation which does the psmx3_memcpy (which internally switches to a memcpy for the same ifdef). Overall, something is off when doing `r[i] = d[i]` with those two datatypes but I'm not sure why. Let me know if you want me to take a different approach!

acgoldma commented:
> @acgoldma This PR is adding atomic verification to the regular fabtests runs. psm3 was failing a lot of them so I've added a few patches to fix most of them. The last patch disables the remaining failing combinations instead of fixing them because I couldn't completely root cause - long double and complex long double reads/swaps are failing and I'm not sure why. Switching the sets to a memcpy of the datatype size also fixes the issue but it takes a performance hit (and also affects all other data types for those ops) so I opted to disable instead of adding that fix. I did this by having all the reads and compares go through the PSM_CUDA | PSM_ONEAPI implementation which does the psmx3_memcpy (which internally switches to a memcpy for the same ifdef). Overall, something is off when doing `r[i] = d[i]` with those two datatypes but I'm not sure why. Let me know if you want me to take a different approach!

We will take a look, but if you have time, it might be better to also adjust psmx3 to use `memcpy()`, they are defined above the changes you made in psmx3_atomic.c

Also, can you give the exact commands to replicate this so we can reproduce internally.

aingerson commented:
@brooksmi Ah shoot, missed adding that. Will add, thank you!

aingerson commented:
@brooksmi Thanks for that context, very helpful. Where do you see that a long double is 8 bits? (Sorry I'm not familiar with long double CPU operations).
The failure I'm seeing is similar to what you're finding in regards to the remaining 48 bits - the verification code fills 128 bits and then checks sizeof(long double) which is 128 bits but psm3 only filled 80 so the check fails. That's why the memcpy fixes it because it does a memcpy(dst, src, sizeof(long double) so it copies all 16 bytes instead of the 10 that the set does. It seems like the verification code needs to only check 10 but I'm wondering how we come to that conclusion if sizeof(long double) returns 16?

aingerson commented:
@brooksmi Yeah, that's what I'm going to move to. It was originally that way but I moved to the memset/memcmp because it was more performant and simplified the implementation for complex types. Back to the original method it is... thanks for your help!

aingerson commented:
@brooksmi Thanks for the context with the long double issues, Michael! I've repushed a version that does an element by element set and check to be explicit with the datatype. I've kept the three psm3 patches though to resolve a few actual psm3 issues regarding complex compares, missing ops, and missing error checks. Let me know if those look ok to you. Thanks!

aingerson commented:
@j-xiong This is finally good to go! I added testing for all the atomics on windows to make sure the implementation was correct on both platforms. Intel CI failure is a known issue Zach is fixing soon and unrelated (MPI rxm IMB slurm timeout). Had to add a few more patches because of psm3 and windows bugs. Review or merge if it looks good to you, thanks!
